packages/components/src/lib/controller/overlay/useOverlayController.ts

```diff
@@ -25,7 +26,26 @@ export const useOverlayController = (
   });
   const controllerFromContext = useOverlayContext()[overlayType];
 
-  return reuseControllerFromContext && controllerFromContext
-    ? controllerFromContext
-    : newController;
+  const controller =
+    reuseControllerFromContext && controllerFromContext
+      ? controllerFromContext
+      : newController;
+
+  useEffect(() => {
+    const disposers: (() => void)[] = [];
+
+    if (onOpen && !controller.isOpen) {
+      disposers.push(controller.addOnOpen(onOpen));
+    }
+
+    if (onClose && controller.isOpen) {
+      disposers.push(controller.addOnClose(onClose));
+    }
+
+    return () => {
+      disposers.forEach((dispose) => dispose());
+    };
+  }, [onOpen, onClose, controller.isOpen]);
```

Comment:
The actual open state should have any effect on binding handlers.

```suggestion
  useEffect(() => {
    const disposers: (() => void)[] = [];

    if (onOpen) {
      disposers.push(controller.addOnOpen(onOpen));
    }

    if (onClose) {
      disposers.push(controller.addOnClose(onClose));
    }

    return () => {
      disposers.forEach((dispose) => dispose());
    };
  }, [onOpen, onClose]);
```